### PR TITLE
[GlobalIsel] Combine G_ADD and G_SUB

### PR DESCRIPTION
**Pull request by tschuett**

### CODE REVIEW

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 1]
```diff
@@ -1634,6 +1634,78 @@ extract_vector_element_shuffle_vector,
...
+
+// fold (A+(B-(C+A))) to (B-C)
+def APlusBMinusCPlusA : GICombineRule<
+   (defs root:$root),
+   (match (G_ADD $add1, $C, $A),
+          (G_SUB $sub1, $B, $add1),
+          (G_ADD $root, $A, $sub1)),
+   (apply (G_SUB $root, $B, $C))>;
+
+def integer_reasso_combines: GICombineGroup<[
```
Comment (arsenm) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1698]: typo reasso->reassoc?

**Review by arsenm — COMMENTED** [review]
All of these probably need a hasOneUse check?

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    %x:_(s32) = COPY $w0
+    %a:_(s32) = COPY $w0
+    %b:_(s32) = COPY $w0
+    %zero:_(s32) = G_CONSTANT i32 0
+    %sub:_(s32) = G_SUB %zero, %a
+    %add:_(s32) = G_ADD %sub, %b
+    $w0 = COPY %add
+    RET_ReallyLR implicit $w0
+
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir:29]: Test the multi-use intermediate cases

**Comment by tschuett** [comment]
There are no hasOneUse checks in the dag. We would need to jump into C++ for the checks.

**Comment by tschuett** [comment]
They are from `visitADDLike`. There are no one-use checks. Fix the DAG or cheap and fast combines.

**Comment by arsenm** [comment]
> There are no hasOneUse checks in the dag. We would need to jump into C++ for the checks.

We should have that? Arguably hasOneUse should be the default behavior for these patterns, you should need to do something to declare another use is OK

**Comment by jayfoad** [comment]
> // fold ((0-A) + B) -> B-A

For cases like this it's not clear that a `hasOneUse` check is needed. If you do the transform when there are multiple uses you'll end up with the same number of operations but less latency, so it seems like a win. Before:
```
  sub x, 0, a
  add y, x, b // dependent on result of sub
  // both x and y are used after this
```
After:
```
  sub x, 0, a
  sub y, b, a // not dependent on first sub
  // both x and y are used after this
```
There _might_ be register pressure effects but they are pretty hard to reason about at the GMIR level.

**Comment by tschuett** [comment]
From `visitADDLike`:
```
// fold ((0-A) + B) -> B-A
if (sd_match(N0, m_Neg(m_Value(A))))
    return DAG.getNode(ISD::SUB, DL, VT, N1, A);
```

File: llvm/include/llvm/Target/GlobalISel/Combine.td [hunk 3]
```diff
@@ -1634,6 +1634,78 @@ extract_vector_element_shuffle_vector,
...
+// fold (A+(B-(C+A))) to (B-C)
+def APlusBMinusCPlusA : GICombineRule<
+   (defs root:$root),
+   (match (G_ADD $add1, $C, $A),
+          (G_SUB $sub1, $B, $add1),
+          (G_ADD $root, $A, $sub1)),
+   (apply (G_SUB $root, $B, $C))>;
+
+def integer_reassoc_combines: GICombineGroup<[
+ZeroMinusAPlusB,
```
Comment (arsenm) [inline llvm/include/llvm/Target/GlobalISel/Combine.td:1699]: Indent these?

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir [hunk 4]
```diff
@@ -0,0 +1,226 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
+# RUN: llc -mtriple aarch64 -run-pass=aarch64-prelegalizer-combiner -verify-machineinstrs %s -o - | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir:2]: Don't think you need -verify-machineinstrs

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    %b:_(s64) = COPY $x1
+    %c:_(s64) = COPY $x2
+    %zero:_(s64) = G_CONSTANT i64 0
+    %add1:_(s64) = G_ADD %c, %a
+    %sub1:_(s64) = G_SUB %b, %add1
+    %add:_(s64) = G_ADD %a, %sub1
+    $x0 = COPY %add
+    RET_ReallyLR implicit $x0
+
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir:223]: Test vectors/

File: llvm/include/llvm/Target/GlobalISel/Combine.td (hunk 3 above)
Comment (tschuett) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1699]: Could please elaborate? Indent by 2 spaces?

Comment (arsenm) [reply llvm/include/llvm/Target/GlobalISel/Combine.td:1699]: Yes

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    %add1:_(<2 x s64>) = G_ADD %c, %a
+    %sub1:_(<2 x s64>) = G_SUB %b, %add1
+    %add:_(<2 x s64>) = G_ADD %a, %sub1
+    $q0 = COPY %add
+    RET_ReallyLR implicit $x0
+
+
+
+
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-integer.mir:257]: Extra newlines at EOF

**Pull request merged**